crates/ra_ide_api/src/goto_definition.rs

```diff
@@ -194,6 +198,29 @@ fn named_target(file_id: FileId, node: &SyntaxNode) -> Option<NavigationTarget>
         .accept(node)
 }
 
+/// Tries to resolve a `path` first using the given `node_resolver`, if that fails to return anything
+/// then it attempts to resolve the path using the given `function` when provided.
+fn try_resolve_path(
+    db: &RootDatabase,
+    path: &hir::Path,
+    node_resolver: &hir::Resolver,
+    function: Option<hir::Function>,
+) -> Option<Resolution> {
+    let resolved = node_resolver.resolve_path(db, path);
```

Comment:
I'm not 100% sure about `try_resolve_path`, it works for the case where the cursor is on the return type of `Self`

```rust
struct Foo;
impl Foo {
    pub fn new() -> Self<|> {
        Self {}
    }
}
```
But there might be a simpler / cleaner way to implement this ?

Reply:
Yeah, I think that maybe `hir::source_binder::resolver_for_node` function should be fixed? I think if the cursor on `Self`, it should return a resolver with generic parameters in scope?

Reply:
Hmm, how would that work ?

Reply:
So, try_resolve_path effectively tries two resolvers: one we get from source binder and one from function. THhat prpbably means that source_binder should have returned the second resolved inthe first place?

Reply:
Something like ?
```rust
//....
} else if let Some(e) = ast::FnDef::cast(node) {
    let module = module_from_child_node(db, file_id, e.syntax())?;
    Some(function_from_module(db, module, e).resolver(db))
}
//....
```